metron-analytics/metron-profiler-common/src/main/java/org/apache/metron/profiler/hbase/DecodableRowKeyBuilder.java

```diff
@@ -0,0 +1,382 @@
+      if (saltLength <= 0 || saltLength > MAX_FIELD_LENGTH) {
+        throw new IllegalArgumentException(String.format("Invalid salt length; max allowed '%d', got '%d'", MAX_FIELD_LENGTH, saltLength));
+      }
+
+      // decode the salt
+      byte[] salt = new byte[saltLength];
+      buffer.get(salt);
+
+      // validate the profile length
+      int profileLength = buffer.getInt();
+      if (profileLength <= 0 || profileLength > MAX_FIELD_LENGTH) {
+        throw new IllegalArgumentException(String.format("Invalid profile length; max allowed '%d', got '%d'", MAX_FIELD_LENGTH, profileLength));
+      }
+
+      // decode the profile name
+      byte[] profileBytes = new byte[profileLength];
+      buffer.get(profileBytes);
+      String profile = new String(profileBytes);
+
+      // validate the entity length
+      int entityLength = buffer.getInt();
+      if (entityLength <= 0 || entityLength > MAX_FIELD_LENGTH) {
+        throw new IllegalArgumentException(String.format("Invalid entity length; max allowed '%d', got '%d'", MAX_FIELD_LENGTH, entityLength));
+      }
+
+      // decode the entity
+      byte[] entityBytes = new byte[entityLength];
+      buffer.get(entityBytes);
+      String entity = new String(entityBytes);
+
+      // decode the groups
+      List<Object> groups = new ArrayList<>();
+      int numberOfGroups = buffer.getInt();
+      for (int i = 0; i < numberOfGroups; i++) {
+
+        // validate the group length
+        int groupLength = buffer.getInt();
+        if (groupLength <= 0 || groupLength > MAX_FIELD_LENGTH) {
+          throw new IllegalArgumentException(String.format("Invalid group length; max allowed '%d', got '%d'", MAX_FIELD_LENGTH, groupLength));
+        }
+
+        // decode the group
+        byte[] groupBytes = new byte[groupLength];
+        buffer.get(groupBytes);
+
+        String group = new String(groupBytes);
+        groups.add(group);
+      }
+
+      // decode the period
+      long periodId = buffer.getLong();
+      long duration = buffer.getLong();
+      ProfilePeriod period = ProfilePeriod.buildFromPeriod(periodId, duration, TimeUnit.MILLISECONDS);
+
+      return new ProfileMeasurement()
+              .withProfileName(profile)
+              .withEntity(entity)
+              .withGroups(groups)
+              .withProfilePeriod(period);
+
+    } catch(BufferUnderflowException e) {
+      throw new IllegalArgumentException("Unable to decode the row key", e);
+    }
+
+  }
+
+  /**
+   * Builds the 'group' component of the row key.
+   * @param groups The groups to include in the row key.
+   */
+  private byte[] encodeGroups(List<Object> groups) {
+
+    // encode each of the groups and determine their size
+    int lengthOfGroups = 0;
+    List<byte[]> groupBytes = new ArrayList<>();
+    for(Object group : groups) {
+      byte[] groupB = Bytes.toBytes(String.valueOf(group));
+      groupBytes.add(groupB);
+      lengthOfGroups += groupB.length;
+    }
```

Comment:
Currently, it prepends each group with its length instead of using a delimiter. There are some unit tests that exercise this.   

Were you thinking this did not work at all or were you suggesting a better way to do it?